### ISSUE TITLE
"Mixed Mode" routing - Cannot make it work

### ISSUE THREAD

**Issue by PhillipHamlyn** [PiranhaCMS/Piranha#475]
I have a typical MVC site that I want to have mixed routing on - I have some fixed pages with coded controllers which I want to use (I've created PageTypes in code for them). I want to use the CMS capability to allow users to create new pages in the site, but cannot get the routing to work.

I've turned PassiveMode to False.

in my RouteConfig I've got;
  routes.MapRoute(
                "Default",                                              // Route name
                "{controller}/{action}/{id}",                           // URL with parameters
                new { controller = "Home", action = "Index", id = "" }  // Parameter defaults
            );

I've tried;

routes.MapRoute(
                name: "Default",
                url: "{controller}/{action}/{id}",
                defaults: new { controller = "Home", action = "Index", id = UrlParameter.Optional },
                namespaces: new[] { "xxx.Controllers" }

My problem is that I cannot access any pages created by the user in the CMS but I can access my hard coded pages (which call out to the CMS via Model Properties).

There is something I'm missing here - can anyone help ?

**Comment by PhillipHamlyn** [comment]
I'm very puzzled about this - when I examine the MVC RouteTable.Routes with PassiveModel set to False I don't see any routing at all for my Piranha pages - only for the authorisation and manager controllers. I'd expect to see a controller added and removed when I switch between PassiveMode True and False. Perhaps this is a clue to why my CMS pages don't get served up ?

**Comment by tidyui** [comment]
Hi there! If you run **passive mode** no CMS-pages will be served by the routing, in your case you should probably stick to having it as **true**.

Now the routing in Piranha CMS is very simple, it runs as a HttpModule that intercepts the request **before** it is passed on to MVC.

1. The request is checked if it matches any permalinks for pages, posts or media
2. If it does the request is simply rewritten to the controller/action route for that page/post type
3. If the request doesn't match any permalink it is passed on to ASP.NET MVC

This is true for everything EXCEPT for the start page. Requests to "~/" will always be served by the start page of the site.

Regards

**Comment by PhillipHamlyn** [comment]
I see that in the source code - but still; with passiveMode = False I get passed through all the requests for my CMS managed pages into my MVC route, which the fail. I am using IIS Express which forces localhost:98749 (random port number) but I guess your just using the localhost portion, so I've "localhost" in the site hostnames. 

In order to try to eliminate the use of "localhost" (which is obviously not ideal) I've tried moving the project to a local IIS and using my machine host name, but I'm getting a flood of SignalR messages from your CMS pages back to the original localhost:98749 location causing cross-site-scripting issues.

Does any of this make any sense to you as a pattern of problems ?

**Comment by serrnovik** [comment]
Hi there! Getting similar issue. I really like the concept, manager and overall functionality of Piranha, but struggling with integrating it.

I want to have partially CMS to integrate to MVC application that has some hard coded functionality, allows administrator to edit parts of some pages and completely manage (create sub pages etc of another). 

So for sake of testing I went with default MVC template. I have there homecontroller with index, about etc. pages. 

If I run application with passiveMode value="false" I'm able to create pages. Custom pages with custom views. But even with <exactPermalinkMatching value="true" /> I can't get any of standart MVC controllers. So for instance I don't have /Home/About in CMS and I get 404 error instead of routing to controller as it should according to post above. 

If I turn passiveMode ON, then (as it should be) I have my controllers working as in any MVC application but permalinks not working. As far as I understood from docs If I want to get content from CMS I should inherit from SinglePageController (HomeController  : Piranha.Mvc.SinglePageController) and edit view to display what I need. I've spent quite some time trying to get the answer how to avoid without result and recive [NullReferenceException: Object reference not set to an instance of an object.]
   Piranha.Models.Page.GetByPermalink(Guid siteTreeId, String permalink, Boolean draft). It doesn't matter if I have permalink with the same name or not looks like Request["permalink"] = '((System.Web.Mvc.Controller)(this)).Request' is null and this breaks in BaseController class. That is probably different question but for the same issue.

In conclusion. Looks like Me and Phillip are solving the same problem. But getting completely different results. With passiveMode false he gets MVC route and I get CMS always. Plus it is not clear how to get pages working in passivemode.

**Comment by tidyui** [comment]
You are all right!!! I finaliy realized what you meant and YES, there is actually a bug in the current version that was a result from a pull request in one of the more frequent versions. The issue was resolved last year and I pushed it to the master branch:

https://github.com/PiranhaCMS/Piranha/commit/e4ee5c581af636f70c1b5309226bd823e1c89b1a

However this doesn't help you much if you install the package from NuGet. Either you get the latest version here from GitHub, or I could put an Alpha up for the upcoming verison. You'll get a bunch a good stuff (check it out at https://github.com/PiranhaCMS/Piranha/milestones/Version%202.2.6) but the new **Blocks** feature isn't 100% finished.

**Comment by serrnovik** [comment]
Thank you for your answer. That is a good news! Having Aplha will be great. If not I'll get master branch and report back soon once I rebuild project manually without nuget it in my scenario.

**Comment by serrnovik** [comment]
I finally had time to check it further. I used master branch and experimented using template site (as I had some issues with clean MVC site probably because I missed some step in moving necessary files and setting all configurations manually)

I have my **problem solved only partially. **

First of all I can confirm that having CMS (with passivmode=false) let me use my controllers side by side. If no permalink with certain address I can have my controller if there is permalink I get what CMS let me see. That is great!

The problem I still have (or don't understand logic) is when I extend controller inheriting from SinglePageController or if I set passivemode=true and try to go by permalink adress. In both of this cases I get NullReffeence error form GetByPermalink method. 
It doesn't even matter what I put into Action methods - that code won't be reached as the issue will be in controller inicialisation. To replicate this behaviour I was following https://piranhacms.org/docs/models/extending-models instructions the moment I inherit from singlepagecontroler all breaks up. 

Is it me missing something or it is a bug? If it is not a bug what is the right way to run custom controller that has it's logic and model to combine it's content with content availible for edit in CMS?

**Comment by tidyui** [comment]
Hi there! The singlepage & singlepost controllers can **only** be used for pages & posts that are routed through the CMS and rewritten internally. Direct accesses to the controller will fail as the base classes tries to retrieve the current permalink from the internal request!

**Comment by serrnovik** [comment]
Thank you for little clarification.

Ok, I tried having say testController and page created in CMS with permalink test/index. I suppose that with controller and permalink having the same route controller should have been called. What is the right way to do achive this goal?

So basicly two questions:

1. What is the way to have cotroller with some custom logic and CMS render page with content from both (as was in example - list products form database and have an ability for content editor to put some contenent on part of this page)?
2. In passive mode. What is the way to render page from CMS?

**Comment by PhillipHamlyn** [comment]
To achieve (1) I just add the pages to the CMS system, and in Razor just use the Piranha HTML Helpers to pull static data from the CMS. I get to choose what content I delegate to the CMS that way. Of course, I'm still stuck in being unable to render a "pure" CMS page that doesn't have any MVC code behind it.

**Comment by serrnovik** [comment]
Hi, thank you Phillip. Sounds like a nice solution. 

I've already tried to find helper. I found two (UI and Site). There are a few very useful such as creating links or render head section. But I can't see a method there that let me, for instance render html section? I even browsed through the sources and didn't find such method. Have not found solution in wiki either.

**Comment by tidyui** [comment]
I can explain how it works tomorrow, I think you're just approaching it the wrong way!

Regards

**Comment by serrnovik** [comment]
Looking forward to it if there is the better way.

**Comment by serrnovik** [comment]
Hi Håkan, so any chance for finding time to put some words for a better solution...

**Comment by PhillipHamlyn** [comment]
Hi,

Did you ever release a fix for Mixed Mode as a pre-release on NuGet ? I can't find it. Sad to say but I'm pretty much going to have to drop this unless I can get it working quite soon. 

Have you got any tracing I can look at in terms of the Http intercepts and decisions made for the routing decisions ? Can I hook any events to try and find out why I just cant open my MVC pages after I install and configure Piranha MVC ? I've tried under IIS Express and full IIS, using localhost and named host and still can't get it working reliably (or at all having removed and reinstalled Piranha). Tought I'd use Glimpse to try and see whats going on, but its killed Glimpse as well :-(

I'll give it one more try by downloading the source code and integrating at that level - at least I'll be able to debug it to see where my page calls are being killed.

**Comment by tidyui** [comment]
I've been away on vacation getting some well deserved time off, but I've uploaded the current beta packages of Piranha CMS to **NuGet**. You can update your existing packages by including pre-releases in the package manager.

**Comment by serrnovik** [comment]
That is great news! Hope vacations was great as well. Could you please write here or somewhere in docs what is best practice to achieve questions 8 posts above?

**Comment by tidyui** [comment]
And here is the **simple** explanation on how the routing & CMS works.

## Page & Post Types
Almost all CMS's out the are based on templates in one way or another. In Piranha this is called **Page Types** and **Post Types**. A type defines the type of content a page of this template can have. For example you could have the page types.

1. One Column Page
2. Two Column Page

The pages would have **one** and **two** html regions. When requests are made for pages created in the CMS of these types the request then has to be **rewritten** to a route that will handle the data and then render it by the correct view. All controllers handling requests **rewritten by the CMS** (this is i active mode) should inherit from either **SinglePageController** or **SinglePostController**

## Adding data from outside the CMS
You could for example have a page displaying some kind of data not controller by the CMS, let say **products**. The correct way of turning this into a Page Type would be to first create a page type called **Product Page**. This page type should contain the content regions needed, but could also contain for example a product category or something else used to filter the products from outside the CMS.

When setting up the model for the ProductPage view you could either **inherit** from PageModel/PostModel, or do a custom model with a property for the CMS data, i.e:

    public class ProductPageModel1 : PageModel {
      public IList<Products> { get; set; }
    }

    public class ProductPageModel2 {
      public IList<Products> { get; set; }
      public PageModel CMSData { get; set; }
    }

Which solution works best is dependent on your application and the amount of code you had before installing Piranha.

## Handling a lot of application pages
In many cases with existing application you already have a vast amount of controllers doing stuff. You might want to add these into the CMS to let content managers control the structure of the site, but adding a Page Type for each different page would not be feasible. A simple solution would then be to create a single Page Type called **ApplicationPage**. This page type could contain some basic content fields, like a HTML region for adding text.

The trick is to on the Page Type set the option `Allow pages to change the route` (you can read about this in section 2.5 at http://piranhacms.org/docs/pages/page-types). This way you could reuse this Page Type for **all** of you application pages, setting the route of the page on the page instance and not cluttering the CMS admin with useless Page Types.

## Passive mode
Passive mode should **ONLY** be used when you have a finished, already completed site and you realize afterwards that you need to allow managers to update content. By enabling passive more **ALL ROUTING** for pages & posts are disabled and only your own controllers will respond to request.

You can then add Page Types as described above and build a site structure containing the content needed to be administrated from the CMS, but you will then need to load the correct PageModel manually in each of your controller/action and add it to your models.

I hope this helps to clarify the setup a bit.

Regards

**Comment by PhillipHamlyn** [comment]
Hi tidyui - I get the theory - I just don't experience it working in practise. 

Heres my experience of working with an existing MVC application; this application is working under IIS Express.

Installed PiranhaCMS.MVC 2.2.6-beta, and altered the RouteHandler as documented. Altered web.config to remove duplicate Authentication tag.

Tried logging in to /manager using known good credentials. Manager page does not log in but just refreshes. Entered wrong credentials - manage page correctly reports "incorrect credentials". Assume inability to login is a beta issue.

Tried opening my Home page (no authentication required - anon access and no [Authorize] directive). Got 404. All my pages report 404.

Switched passiveMode = true (in desperation). My pages now render, but of course, since there is no server side URL rewrite then I would not be able to include CMS pages alongside my hard coded pages.

What I desperately need is for you to have a logging capability so I can actually trace the re-write decisions the platform is actually making, otherwise were just going to be stuck in this loop of you telling me "it works" and me saying "no it doesn't". Don't get me wrong, its a nice platform - but I just can't make it work as advertised currently. It seems the Achilles heel of the underlying design is the auto-magic url rewriting (why not use MVC routes on your MVC version - its what they were designed for :-) ) and that its not debuggable or diagnosable.

**Comment by tidyui** [comment]
So, I've created a new repo here where I've added an MVC example for **Mixed Mode**. In my example I've assumed that you've added the package `piranha.mvc` from NuGet @PhillipHamlyn.

https://github.com/PiranhaCMS/Examples

I followed the guidlines from *Adding Piranha CMS to an existing application* at http://piranhacms.org/docs/introduction/getting-started *except* for letting `passiveMode` be **false**.

For me the setup works perfectly exept for the fact that CMS generated pages doesn't work. The reason for this is that the template package `piranhacmsmvc` bundles two **controllers** and **views** for rendering CMS content. These are of course not present in the lib setup of `piranha.mvc`.

In my example repo I've added the default controllers & views for pages & posts.

**Comment by PhillipHamlyn** [comment]
Thanks for making the repo - I've downloaded it and linked against my CMS DB and yet :-) it works. 

I found the issue with logging in to the manager and perhaps a clue to my woes. If I set the web application settings to SSL Enabled and then access the site via Https then no manager or other pages are rendered. My sites are always configured as Https. I never thought to turn that off.

I reconfigured my site to use Http and now I can now load the management pages (horay !). All my controllers are marked with [Authorize][RequireHttps] which may be causing the issue. Have you been able to make your repo work over Https ?

**Comment by tidyui** [comment]
Yes I could definitely look into https, I wasn't aware there was an issue with it and no one's reported it :)

I didn't quite follow your first paragraph though, did the example repo work for you?

**Comment by tidyui** [comment]
Because if the test repo works I'm closing this issue and marking it for release with 2.2.6 and then I suggest we open a new issue for https!

**Comment by PhillipHamlyn** [comment]
Yes - Ok - the repo worked for me Http only. Thanks for your help. I will temporarily strip out my [Authorize][RequireHttps] and see if I get further.
